docs/source/cpp/gandiva.rst

```diff
@@ -0,0 +1,151 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+.. cpp:namespace:: arrow::compute
+
+===============================
+The Gandiva Expression Compiler
+===============================
+
+Gandiva is a runtime expression compiler that uses `LLVM`_ to generate
+efficient native code for projections and filters on Arrow record batches.
+Gandiva only handles projections and filters. For other transformations, see
+:ref:`Compute Functions <compute-cpp>`.
+
+Gandiva was designed to take advantage of the Arrow memory format and modern
+hardware. Compiling expressions using LLVM allows the execution to be optimized
+to the local runtime environment and hardware, including available SIMD
+instructions. To reduce optimization overhead, many Gandiva functions are
+pre-compiled into LLVM IR (intermediate representation).
+
+.. _LLVM: https://llvm.org/
+
+
+Building Expressions
+====================
+
+Gandiva provides a general expression representation where expressions are
+represented by a tree of nodes. The expression trees are built using
```

Comment:
It might be worth explaining what the nodes represent in particular. While I can figure it out in the next section, where function nodes are discussed, having to wait for that clarity feels weird to me.

Reply:
I guess what would really be helpful would be to draw an expression tree, but idk if I want to do that in this PR.

I don't really know how to describe what a node itself means. That's why I first describe the leaves, and then the composite nodes. And then give an example.

If you have a suggestion, though, I'd happily take it.

Reply:
I think something's off in the writing, then -- I did not find myself thinking of "composite nodes" after reading, only leaf nodes, then something that's above that. It might be worth going more in depth in an additional paragraph, at least. 

As for how to describe them, internal nodes represent operations that handle multiple leaves, while leaves contain fields and literals, I believe? I haven't written any Gandiva, so if I'm off-key here, we could talk further.